pkg/kubelet/kubelet_getters.go

```diff
@@ -274,6 +274,25 @@ func (kl *Kubelet) getPodVolumePathListFromDisk(podUID types.UID) ([]string, err
 	return volumes, nil
 }
 
+func (kl *Kubelet) getMountedVolumePathListFromDisk(podUID types.UID) ([]string, error) {
+	mountedVolumes := []string{}
+	volumePaths, err := kl.getPodVolumePathListFromDisk(podUID)
+	if err != nil {
+		return volumePaths, err
+	}
+	glog.V(4).Infof("Volume paths exist for pod %s: %v", podUID, volumePaths)
+	for _, volumePath := range volumePaths {
+		isNotMount, err := kl.mounter.IsLikelyNotMountPoint(volumePath)
```

Comment:
I do not like check for mount points here. Flex volume may not use mount() at all, it may populate the directory in another way - e.g. Secret-like volume. So it may be "used" and TearDown needs to be called while the directory is not in /proc/mounts.

IMO, the presence of the directory is enough to call TearDown() of particular volume plugin and the directory should be removed only when it succeeds.

Of course, if we can't reconstruct the volume, unmount + rmdir is the best we can do, but as a last resort.

Reply:
@jsafrane given the lateness of this work in 1.10 and our pending release, can you advise a solution, or clarify the risk of implementing this way?  In short, is this concern a blocker for approval?

Reply:
I am not sure if this is a blocker, if what flexvolume did was populated a directory (not a mount) and it is in "use", then surely a running pod will mean volume gets added to ASW and `Teardown` will be called when pod is deleted.

The worst that can happen is - when a pod using flexvolume (and assuming flexvolume did not create mount points but just populated plain directories) is deleted while kubelet was down and the race condition in question prevented volume from being added to ASW then pod may be deleted without calling `Teardown` on left over directories. 

Lets review what happens in 1.9 for flexvolume that does not create mount points. Such volumes will not be processed by reconstruction process - https://github.com/kubernetes/kubernetes/blob/release-1.9/pkg/kubelet/volumemanager/reconciler/reconciler.go#L490 and hence in similar fashion as this PR will leave directories on which `Teardown` isn't called.

Reply:
@jdumars I agree with @gnufied, my comment is not blocking 1.10. Flex volume that does not mount + kubelet restart + forced pod deletion need to meet at the right time to leave an orphan directory. We can fix it in 1.10.1 or 1.11.

Reply:
The purpose of this checking is to block pod deleted from API server if there are still some directories and mounts left behind. In normal case, it does not interfere with reconciler tear down process. By normal case, I mean the following workflow
1. Pod is created and added in desired state, and reconciler mounts volumes and add to actual state.
2. User tries to delete pod, so pod deletion timestamp is set
3. Kubelet kills containers
4. Desired state populator removes the pod from desired state (it checks if pod is in terminating state, containers are killed and actual state has the volume already)
5. Reconciler issues tear down volume process.

In case of pod is deleted during kubelet restart, the following will happen.
1. Pod (in terminating state) is added back to desired state first.
2. Reconstruction will not proceed if volume in desired state
3. Reconciler tries to set up volume and add it back the actual state. In the mean time, kubelet tries to delete the pod. With the PR, it will fail to delete if directories exist and mounted. So reconciler have a chance to make the volume into the actual state. 
4. Desired state populator can now remove volume since actual state has the volume. So reconciler will tear down the volume. After that kubelet can delete pod.

In the case of directories are not mounted, race condition might happen like the following, kubelet deletes the pod so desired state populator removes the pod so reconciler will not issue set up volume and add it to the actual state. But it should be a rare case. Before 1.10, we have quite some volume plugins cannot reconstruct volume correctly so they are not cleaned up. With the new design, we give more time for reconciler to reconstruct the state from desired state by delaying the deletion of pod from desired state. So overall, it should improve the situation, I think.

Reply:
yes, the overall situation is better than before this PR.